### DISCUSSION TITLE
RFC: Better Syntax for the Borrow Statement

### DISCUSSION THREAD

**Discussion by eudoxia0** [austral/austral#479] [Ideas]
The current borrow statement looks like this:

```austral
borrow x as xref in R do
    -- etc.
end;
```

And the mutable version:

```austral
borrow! x as xref in R do
    -- etc.
end;
```

This has some drawbacks:

1. `borrow` is too long.
2. The difference between the two is a tiny, easily missed exclamation mark.
3. Unlike everywhere else in the language, there is no type annotation for `xref`, you have to mentally reconstruct it by looking up the type of `x`.
4. Because `borrow` is too long, in the interest of convenience, this is the only statement that doesn't end with the name of the statement. We have `end if`, `end case`, `end for`, and `end while`, but `end borrow` is just `end`.
5. Also there is no statement version of reborrow.

There must be better ways.

# "Scoped Let" statements.

So the idea I currently have is to make the borrow statement into a "scoped let", where the expression on the right hand side of the assignment tells you what it's doing.

Example:

```
let xref: &[T, R] := &x do
    -- etc.
end let;
```

This is a equivalent to a read-only `borrow` statement.

1. There is a type annotation.
2. The type annotation and the `&x` expression tell you you're taking a read-only reference.
3. The name of the region is introduced in the type annotation.
4. The statement ends with the name of the statement.
5. `end let` doesn't take too long to write.

Analogous versions for mutable borrows and reborrows:

```
let xref: &![T, R] := &!x do
    -- etc.
end let;

let xref: &[T, S] := reborrow(x) do
    -- etc.
end let;
```

**Comment by crabmusket** [comment]
This doesn't seem to solve problem 2 (the difference between the two is a tiny, easily missed exclamation mark), but at least it mirrors [existing reference syntax](https://austral-lang.org/spec/spec.html#write-ref). I assume the following would also be valid syntax?

```
let xref: WriteReference[T, R] := &!x do
    -- etc.
end let;
```

**Comment by eudoxia0** [comment]
>the difference between the two is a tiny, easily missed exclamation mark),

Well yes but the exclamation mark exists in two places.

>I assume the following would also be valid syntax?

No, I dropped `Reference` and `WriteReference` for `&` and `&!` respectively. They took too long to write.
